### PR TITLE
[feat] Added differentiable Hausdorff Distance (HD) loss

### PR DESCRIPTION
**Pull request by shijianjian**
### Description
Added differentiable erosion-based HD loss.

### CODE REVIEW

File: kornia/losses/hausdorff.py [hunk 1]
```diff
@@ -0,0 +1,157 @@
...
+class HausdorffERLoss(_HausdorffERLossBase):
+    """Binary Hausdorff loss based on morphological erosion.
+
+    This is an Hausdorff Distance (HD) Loss that based on morphological erosion,which provided
+    a differentiable approximation of Hausdorff distance as stated in :cite:`karimi2019reducing`.
+    The code is refactored on top of `here <https://github.com/PatRyg99/HausdorffLoss/
+        blob/master/hausdorff_loss.py>`__.
+
+    Args:
+        alpha: controls the erosion rate in each iteration. Default: 2.0.
```
Comment (edgarriba) [inline kornia/losses/hausdorff.py:90]:
```suggestion
        alpha: controls the erosion rate in each iteration.
```

File: kornia/losses/hausdorff.py [hunk 2]
```diff
@@ -0,0 +1,157 @@
...
+        erosions: the number of iterations of erosion. Default: 10.
```
Comment (edgarriba) [inline kornia/losses/hausdorff.py:91]:
```suggestion
        erosions: the number of iterations of erosion.
```

File: kornia/losses/hausdorff.py [hunk 3]
```diff
@@ -0,0 +1,157 @@
...
+    Examples:
+        >>> hdloss = HausdorffERLoss()
+        >>> input = torch.randn(5, 3, 20, 20)
+        >>> target = torch.randn(5, 1, 20, 20).long() * 3
+        >>> res = hdloss(input, target)
+    """
+    conv = torch.conv2d
+    avg_pool = nn.AdaptiveAvgPool2d(1)
+
+    def get_kernel(self) -> None:
```
Comment (edgarriba) [inline kornia/losses/hausdorff.py:102]:
```suggestion
    def get_kernel(self) -> torch.Tensor:
```

File: kornia/losses/hausdorff.py [hunk 4]
```diff
@@ -0,0 +1,157 @@
...
+            # image-wise differences for 2D images
+            erosion_max = self.avg_pool(erosion)
+            erosion_min = - self.avg_pool(- erosion)
+            # No normalization needed if `max - min = 0`
+            to_norm = (erosion_max.squeeze() - erosion_min.squeeze()) != 0
+            erosion[to_norm] = (erosion[to_norm] - erosion_min[to_norm]) / (
+                erosion_max[to_norm] - erosion_min[to_norm])
+
+            # save erosion and add to loss
+            eroted = eroted + erosion * (k + 1) ** self.alpha
```
Comment (edgarriba) [inline kornia/losses/hausdorff.py:57]:
```suggestion
            eroded = eroded + erosion * (k + 1) ** self.alpha
```

File: kornia/losses/hausdorff.py [hunk 5]
```diff
@@ -0,0 +1,157 @@
...
+            bound = erosion
+
+        return eroted
```
Comment (edgarriba) [inline kornia/losses/hausdorff.py:60]:
```suggestion
        return eroded
```

File: kornia/losses/hausdorff.py [hunk 6]
```diff
@@ -0,0 +1,157 @@
...
+
+    def forward(self, pred: torch.Tensor, target: torch.Tensor) -> torch.Tensor:
+        """Compute Hausdorff loss.
+
+        Args:
+            pred: predicted tensor with a shape of :math:`(B, C, H, W)` or :math:`(B, C, D, H, W)`.
+                Each channel is as binary as: 1 -> fg, 0 -> bg.
+            target: target tensor with a shape of :math:`(B, 1, H, W)` or :math:`(B, C, D, H, W)`.
+        """
+        assert pred.shape[2:] == target.shape[2:] and pred.size(0) == target.size(0) and target.size(1) == 1, (
```
Comment (edgarriba) [inline kornia/losses/hausdorff.py:70]: add if / raise - same for the other assertion
https://deepsource.io/gh/kornia/kornia/run/f0e4fe5a-a07d-407b-a7c4-6e7b40f37f37/python/BAN-B101

File: kornia/losses/hausdorff.py [hunk 7]
```diff
@@ -0,0 +1,157 @@
...
+        kernel = cross * 0.2
+        return kernel[None]
+
+    def forward(self, pred: torch.Tensor, target: torch.Tensor) -> torch.Tensor:
+        """Compute Hausdorff loss.
+
+        Args:
+            pred: predicted tensor with a shape of :math:`(B, C, H, W)`.
+                Each channel is as binary as: 1 -> fg, 0 -> bg.
+            target: target tensor with a shape of :math:`(B, 1, H, W)`.
```
Comment (edgarriba) [inline kornia/losses/hausdorff.py:113]: add the return description

File: kornia/losses/hausdorff.py [hunk 8]
```diff
@@ -0,0 +1,157 @@
...
+class HausdorffERLoss3D(_HausdorffERLossBase):
+    """Binary 3D Hausdorff loss based on morphological erosion.
+
+    This is a 3D Hausdorff Distance (HD) Loss that based on morphological erosion,which provided
+    a differentiable approximation of Hausdorff distance as stated in :cite:`karimi2019reducing`.
+    The code is refactored on top of `here <https://github.com/PatRyg99/HausdorffLoss/
+        blob/master/hausdorff_loss.py>`__.
+
+    Args:
+        alpha: controls the erosion rate in each iteration. Default: 2.0.
```
Comment (edgarriba) [inline kornia/losses/hausdorff.py:128]:
```suggestion
        alpha: controls the erosion rate in each iteration.
```

File: kornia/losses/hausdorff.py [hunk 9]
```diff
@@ -0,0 +1,157 @@
...
+        erosions: the number of iterations of erosion. Default: 10.
```
Comment (edgarriba) [inline kornia/losses/hausdorff.py:129]:
```suggestion
        erosions: the number of iterations of erosion.
```

File: kornia/losses/hausdorff.py [hunk 10]
```diff
@@ -0,0 +1,157 @@
...
+        >>> hdloss = HausdorffERLoss3D()
+        >>> input = torch.randn(5, 3, 20, 20, 20)
+        >>> target = torch.randn(5, 1, 20, 20, 20).long() * 3
+        >>> res = hdloss(input, target)
+    """
+
+    conv = torch.conv3d
+    avg_pool = nn.AdaptiveAvgPool3d(1)
+
+    def get_kernel(self) -> None:
```
Comment (edgarriba) [inline kornia/losses/hausdorff.py:141]:
```suggestion
    def get_kernel(self) -> torch.Tensor:
```

File: kornia/losses/hausdorff.py [hunk 11]
```diff
@@ -0,0 +1,157 @@
...
+        kernel = torch.stack([bound, cross, bound], dim=1) * (1 / 7)
+        return kernel[None]
+
+    def forward(self, pred: torch.Tensor, target: torch.Tensor) -> torch.Tensor:
+        """Compute 3D Hausdorff loss.
+
+        Args:
+            pred: predicted tensor with a shape of :math:`(B, C, D, H, W)`.
+                Each channel is as binary as: 1 -> fg, 0 -> bg.
+            target: target tensor with a shape of :math:`(B, 1, D, H, W)`.
```
Comment (edgarriba) [inline kornia/losses/hausdorff.py:154]: add the return description

**Review by edgarriba — CHANGES_REQUESTED** [review]
LGTM - add tests, attach to docs and update changelog

**Comment by shijianjian** [comment]
> Is there any paper reference that we can add here ?

It has been there in the first place.

**Comment by shijianjian** [comment]
@ducha-aiki There are [smoke test errors](https://github.com/kornia/kornia/pull/1254/checks?check_run_id=3580607078) in LoFTR, can you take a look?

**Review by edgarriba — APPROVED** [review]
LGTM. Are you aware of any potential example that we could show for this functionality ?

**Comment by shijianjian** [comment]
> LGTM. Are you aware of any potential example that we could show for this functionality ?

This is used to improve semantic segmentation performances, similar to dice loss. Probably worth adding some illustrations to the docs.
